dev/core#6208: Prevent membership renewal or type change without completed contribution

rbaugh opened the pull request:
Overview
----------------------------------------
When an existing membership exists and a contribution is attempted, the existing membership either renews or changes type regardless of whether the contribution is completed. This makes sure the contribution is completed before renewing or changing the membership to another type.

Before
----------------------------------------
Membership would renew or change type when the contribution record was created and before even attempting a payment.

After
----------------------------------------
Waits for the Contribution to be completed.

Comments
----------------------------------------
Renewal/type change was moved to the OrderCompleteSubscriber, but not sure what changes or notices we would need to make since this Subscriber could be overwritten in an extension and fail to work as intended.

rbaugh committed:
Initial changes

rbaugh commented on CRM/Contribute/Form/Contribution/Confirm.php:
```diff
@@ -2862,6 +2864,26 @@ protected static function legacyProcessMembership($contactID, $membershipTypeID,
       }
     }
+    elseif ($currentMembership && $contribution->contribution_status_id !== $completedStatusId) {
+      // We have an existing membership and payment isn't complete, we just need to record the contribution
```
The intent is that if there is an existing membership, we are not changing the membership type or renewing it until the contribution is completed. With the current code, you don't even have to really attempt a payment or you can just enter fake CC info that will cause the contribution payment to fail, and the membership would still advance or change types.

What I am doing here is just connecting the line items with the membership so that in the OrderCompleteSubscriber, we can look at the line items and see if we need to process a renewal or change the membership type. Unless this is a free contribution, the payment hasn't been completed when this function is called. I am not sure if at this point a contribution could be completed, but I am just adding in a check to make sure it isn't completed yet and setting things up so we can check later in the OrderCompleteSubscriber.

eileenmcnaughton replied:
@rbaugh so if the action is to alter existing line items to add the membership_id? Or create line items attached to the membership (Ideally if we are altering line items we should do that directly with the LineItem api rather than going through the membership BAO craziness)

rbaugh replied:
The intent is just to make sure the line items are able to be connected with the existing membership so that it can be processed after the order is completed. Not necessarily to change the line items, but just to associate the line items with the membership.

When testing, without this connection here, on the OrderCompleteSubscriber, if you try and locate the line item for the membership, it didn't exist. So with this change, we can check what line items were connected for each contact/line item post payment to then renew it or change the type.

rbaugh replied:
I can change the `$preChangeMembership` variable to something like `$membershipLineItem` in the [updateMembershipBasedOnCompletionOfContribution](https://github.com/civicrm/civicrm-core/pull/34342/files#diff-a0d2012b84f0d6778adc733a16e9c51c5b4cce8206a32a0fe3cdc0d366db31b9R85) to make that more specific as it isn't a look back anymore, just what is the line item and what do we need to do with it.

eileenmcnaughton commented:
I'm trying to think how to replicate this in a unit test

rbaugh commented:
> I'm trying to think how to replicate this in a unit test

There are a few different test I think that could be created. Membership would need to be created first of course.

1. Contribution created with payment offsite, so essentially it should just create the contribution in a pending state.
2. Contribution created with failed payment. 
3. Contribution created with successful payment.

For #1 and #2, the only thing that should happen is the contribution is created, one pending one failed. Nothing should be changed for the membership. For #3, the contribution is created, and completed, along with the membership being renewed or type changed based on the line items of the contribution. #1 might be a little harder to create based on the flow of the payment provider. 

I can assist with some unit tests, I was just trying to make sure this was ok as your previous comment on the issue made it sound like there was more to this. Also, I wasn’t sure what might need to be done since the renewal/type change would now happen in the OrderCompleteSubscriber rather than in the Confirm and then using the MembershipLog to figure out what happened. Is it just notes in the release to make sure people check any potential changes they might have made by creating their own subscriber?? I know for myself we have a client that will need to have a custom subscriber adjusted as the logic wouldn’t work anymore.

rbaugh committed:
Added UnitTests and some fixes

rbaugh commented:
retest this please

eileenmcnaughton commented:
@rbaugh those tests look pretty good! 

Note that the PR queue tests have problems at the moment - it's been the past few days - maybe disk space?

rbaugh commented:
> @rbaugh just in terms of timing - I'm keen to get this fixed 'right' - which might mean it will drag on a bit - I hope that is OK...

I have a patch in on the OrderCompleteSubscriber where this became an issue, so I am ok other than just wanted to get this fixed, as it is larger than just my clients. I'm happy to jump on a call at some point if it feels like we can make more progress that way.

eileenmcnaughton commented:
@rbaugh ok - so this is a regression?

rbaugh commented:
@eileenmcnaughton It’s more of a bug. We had come across it before, but thought it was just something odd with the client. I was then testing for something else and realized the problem. It seems this has been an issue for quite some time, but don’t know how long. So not a recent regression. Just that this basically can allow someone to get free membership upgrades or renewals without this or a custom fix.

rbaugh commented:
@eileenmcnaughton @mattwire 

I am not sure if this section of code would need to be moved to the `OrderCompleteSubscriber` as it currently only seems to exist in the legacyProcessMembership. I am also not quite sure that if at this point you could have a `Completed` Contribution or not.
https://github.com/civicrm/civicrm-core/blob/d6127070d6a6483c0d2bffbf107f6fb81e074331/CRM/Contribute/Form/Contribution/Confirm.php#L2701-L2765

I think that the creation of a Membership is also still needed as we want to create the `Pending` membership before completing or attempting the payment.
https://github.com/civicrm/civicrm-core/blob/d6127070d6a6483c0d2bffbf107f6fb81e074331/CRM/Contribute/Form/Contribution/Confirm.php#L2825-L2864

The main thing that this PR is doing is that it makes sure that if there is an existing membership, that nothing is done to it unless the `Contribution` is marked as `Completed`.

I am not as familiar with this function, but as I am reading through it, it still seems like it is needed. The first section, L2701-L2765 could potentially get moved to the `OrderCompleteSubscriber` as well since it would now be checking for an existing membership and if the `Contribution` was marked as `Completed`. The `OrderCompleteSubscriber` seems like the logical place for this block to go, but I wasn't sure about that much of a code overhaul and I wasn't sure if it could impact other things.

So any help or suggestions would be appreciated in order to see how we can move this forward.

rbaugh committed:
legacyProcessMembership cleanup

rbaugh commented:
@eileenmcnaughton @mattwire 

I have reworked the `legacyProcessMembership` function to remove some of the bloat. Overall, if there is a current membership, then I am just associating the line items with the existing membership and then allowing the rest of the process to be handled in `OrderCompleteSubscriber`. If this is going to be a new membership, then I am still processing the new membership. I switched it to use API4 for the membership create, instead of the BAO, but can move it to API3 if there is an need. I know there were some issues the creep up in API4 that didn't exist in API3, but I am not sure if those have been resolved now. This also associates the new membership with the contribution line item so that once the Contribution is complete, it can be finished in the `OrderCompleteSubscriber` as well.

I tried to add some params that were in the `legacyProcessMembership` function in `OrderCompleteSubscriber` where appropriate. All existing UnitTest, and my new tests, still seem to process just fine. I also ran some renewals, type changes, and new memberships and they seem to be working ok. But these are basic tests and don't account for various oddities that can be created in a contribution.

mattwire commented:
@rbaugh This looks good to me - can you squash the commits down a bit and then I think we should be good to merge

eileenmcnaughton commented on CRM/Contribute/Form/Contribution/Confirm.php:
```diff
@@ -2866,14 +2763,50 @@ private function legacyProcessMembership($contactID, $membershipTypeID, $changeT
+      }
+    }
+
+    CRM_Price_BAO_LineItem::processPriceSet(
```
PLEASE do not call this function - just save the line items using the api. If you have to call this function to get the financial items right then I take back my comment about the above being non-blocking....

eileenmcnaughton replied:
I've added a PR to address the fact this was not clearly deprecated https://github.com/civicrm/civicrm-core/pull/34489

Note that given this is for the NEW membership flow the Order->payment flow should be possible

eileenmcnaughton commented:
@mattwire @rbaugh I think once this is switched to using the order api then `associateMembershipLineItems()` can be deleted again as that is handled via Order Api

pradpnayak commented:
Hi @eileenmcnaughton / @mattwire 
This change appears to have introduced a regression affecting several of our clients. We are currently reviewing additional sites to understand the full extent of the impact, but the following scenario is quite common and has been reproduced consistently since upgrading to CiviCRM 6.13.3.

### Scenario

A membership is configured for auto-renewal using Stripe or Smart Debit.

At some point, either:

* The member requests a membership type change (e.g. via a phone call or request form), or
* The organisation restructures its membership offerings and members are migrated to a different membership type.

In these cases, a staff member manually updates the membership type, either individually through the UI or in bulk using SearchKit or a CSV import.

After the next recurring payment is collected, the membership type is automatically reverted back to the previous membership type associated with the recurring contribution.

For some organisations, this has significant consequences:

* Members may lose access to content they should still be entitled to access.
* Members may gain access to content for which they are no longer eligible.
* Membership reporting and segmentation become inaccurate.

This behaviour did not occur prior to 6.13.3.

### Questions

I am trying to understand the intention behind the changes introduced in this PR and whether this behaviour is expected.

From reviewing the code, it appears that preserving the new membership type would require updating either:

* The Price Field Value ID on the most recent contribution line item to match the new membership type, or
* The template contribution (where one exists).

However, both approaches raise concerns:

* Updating the template contribution line item seems reasonable in principle, but approximately 99% of recurring contributions on our client sites do not have a template contribution.
* Updating the line item on the most recent contribution does not seem appropriate, as that contribution was historically paid for the original membership type and should remain an accurate record of what was purchased at that time.

### Clarification Requested

Could you please confirm whether this is an intentional change?

Specifically:

* Is there now a rule that membership types should not be manually changed when they are linked to a recurring contribution?
* If so, could you point me to any documentation describing this behaviour so that we can advise our clients accordingly and explain any new restrictions or required processes?

Alternatively, if this is not intended behaviour and is considered a bug, then it seems we need a mechanism to preserve membership type changes made through backend processes.

Possible approaches might include:

* Automatically creating or updating a template contribution when a membership type is changed.
* Providing a supported API or workflow for changing membership types on memberships linked to recurring contributions.
* Updating recurring contribution metadata when membership type changes occur.

I can also confirm that the same issue can be reproduced on the official demo site:

https://smaster.demo.civicrm.org/

<img width="1785" height="1060" alt="MembeType" src="https://github.com/user-attachments/assets/e675ccbd-1fed-49ab-9c7a-d5040d3333cd" />


Any clarification on the intended behaviour and recommended workflow would be greatly appreciated.

eileenmcnaughton commented:
@pradpnayak I think you are talking about a regression rather than an intentional behaviour change - I would like to get @rbaugh & @mattwire to weigh in as I was more involved in the choice of functions than in the desired outcome

mattwire commented:
This sounds like a regression. Before the changes were made a membership type would change immediately and get access to that new membership level even if they didn't actually make a payment / payment failed. Following the changes, the membership type only changes once the payment completes.

I think that "Automatically creating or updating a template contribution when a membership type is changed." would be the right approach because we have:
- Recur + Paid Contribution + lineitems = "The original/unchanged state"
- Manual membership type change = "Not recorded anywhere in payment information".
- Recur + Template Contribution + Template lineitems = "The expected state going forward".

If the template contribution does not exist then the latest existing contribution will be used.

So to fix this I think we need to:
1. Identify which codepaths are used for a manual membership type change.
2. Find a common point where we can check if the membership is linked to a recurring contribution.
3. If linked to a recur, create/update a template contribution and change the lineitem to the new membership type.

rbaugh commented:
This PR was intended to make sure that the membership didn't advance without a payment. The bug was that you could submit a contribution page without valid credentials and when the payment failed, it would still advance the membership and therefore grant the membership renewal without the payment. Same thing for a membership type change.

It was also an issue for payment processors where the payment was handled on the processor's website. The Contribution page would renew/change the membership, then send the user to the processor's website. So the user didn't even have to attempt a payment at all.

As @mattwire mentioned, the scenario presented, is that the membership type is being changed and not being reflected on the ContributionRecur, and when someone goes in to renew, the renewal is based on the previous Contribution and then become another membership type change since the two don't match.

I believe the key thing Matt mentioned was "Identify which codepaths are used for a manual membership type change" and how to prevent the bug this PR addressed from becoming a regression itself.

rbaugh commented:
I guess another question to ponder is if the type is changed manually, should the ContributionRecur be updated with a potential price change as well? Or is the recur keeping the same cost yet the member now has a higher priced membership.
